File: CanIWin.cpp
/*
* Time Limit Exceeded
*/
#include<iostream>
#include<unordered_set>

using namespace std;

class Solution {
public:
    bool canIWin(int maxChoosableInteger, int desiredTotal) {
    	unordered_set<int> choosed;
        for(int i = maxChoosableInteger; i > 0; i--) {
        	if(i >= desiredTotal) return true;
        	choosed.insert(i);
        	if(!canIWin(maxChoosableInteger, desiredTotal, i, choosed)) return true;
        	choosed.erase(i);
		}
		
		return false;
    }
    
    bool canIWin(int maxChoosableInteger, int desiredTotal, int total, unordered_set<int>& choosed) {
    	for(int i = maxChoosableInteger; i > 0; i--) {
    		if(choosed.find(i) == choosed.end()) {
    			if(total + i >= desiredTotal) return true;
    			choosed.insert(i);
    			bool next = canIWin(maxChoosableInteger, desiredTotal, total + i, choosed);
    			choosed.erase(i);
    			if(!next) return true;
			}
		}
		return false;
	}
};

int main() {
	
	int maxChoosableInteger, desiredTotal;
	cin>>maxChoosableInteger>>desiredTotal;
	
	Solution *solution = new Solution();
	solution->canIWin(maxChoosableInteger, desiredTotal) ? cout<<"true" : cout<<"false";
	
	return 0;
	
}

File: CanIWin.py
class Solution:
    def canIWin(self, maxChoosableInteger: int, desiredTotal: int) -> bool:
        def canIWin(total, used):
            if used in m:
                return m[used]
            for i in range(maxChoosableInteger, 0, -1):
                cur = 1 << (i - 1)
                if not (used & cur):
                    if i >= total or not canIWin(total - i, used | cur):
                        m[used] = True
                        return True
            m[used] = False
            return False
        if maxChoosableInteger >= desiredTotal:
            return True
        if (1 + maxChoosableInteger) * maxChoosableInteger // 2 < desiredTotal:
            return False
        m = {}
        return canIWin(desiredTotal, 0)


if __name__ == '__main__':
    maxChoosableInteger = int(input())
    desiredTotal = int(input())
    solution = Solution()
    print(solution.canIWin(maxChoosableInteger, desiredTotal))
